commit 3d7ba52b68e4dc5c4d3eb19de436c66ed9bb2f0d
Author: Andreas Schwab <schwab@redhat.com>
Date:   Thu Nov 3 14:26:38 2011 +0100

    Don't fail in makedb if SELinux is disabled

*** a/ChangeLog	Wed Nov 30 12:38:59 2011
--- b/ChangeLog	Wed Nov 30 12:39:17 2011
***************
*** 9,14 ****
--- 9,19 ----
  	(reused_arena): Don't check arena limit.
  	(arena_get2): Atomically check arena limit.
  
+ 2011-11-03  Andreas Schwab  <schwab@redhat.com>
+ 
+ 	* nss/makedb.c (set_file_creation_context): Do nothing if SELinux
+ 	is disabled.
+ 
  2011-10-19  Andreas Schwab  <schwab@redhat.com>
  
  	* sysdeps/x86_64/fpu/math_private.h (libc_feupdateenv): Use
diff --git a/nss/makedb.c b/nss/makedb.c
index 8cee92f..1b19966 100644
--- a/nss/makedb.c
+++ b/nss/makedb.c
@@ -842,7 +842,7 @@ set_file_creation_context (const char *outname, mode_t mode)
 
   /* Check if SELinux is enabled, and remember. */
   if (enabled == 0)
-    enabled = is_selinux_enabled ();
+    enabled = is_selinux_enabled () ? 1 : -1;
   if (enabled < 0)
     return;
 
