Index: glibc-2.23-69-gf5e753c8/localedata/locales/an_ES
===================================================================
--- glibc-2.23-69-gf5e753c8.orig/localedata/locales/an_ES
+++ glibc-2.23-69-gf5e753c8/localedata/locales/an_ES
@@ -134,7 +134,7 @@ postal_fmt    "<U0025><U0066><U0025><U00
 country_ab2   "<U0045><U0053>"
 country_ab3   "<U0045><U0053><U0050>"
 country_num   724
-lang_ab       "<U0041><U004E>"
+lang_ab       "<U0061><U006E>"
 % E
 country_car    "<U0045>"
 % aragonés
