commit 850fb039cec802072f70ed9763927881bbbf639c
Author: Andreas Schwab <schwab@linux-m68k.org>
Date:   Tue Dec 6 11:10:06 2011 +0100

    Fix a wrong constant in powerpc hypot implementation

--- a/ChangeLog	2011-12-05 20:28:47.000000000 -0700
+++ b/ChangeLog	2011-12-07 13:14:29.619858242 -0700
@@ -1,3 +1,7 @@
+2011-12-06  Andreas Schwab  <schwab@linux-m68k.org>
+
+	* sysdeps/powerpc/fpu/e_hypot.c (twoM600): Correct value.
+
 2011-12-02  Andreas Schwab  <schwab@redhat.com>
 
 	* misc/bits/select2.h (__FD_ELT): Mark as extension.  Add
diff --git a/sysdeps/powerpc/fpu/e_hypot.c b/sysdeps/powerpc/fpu/e_hypot.c
index a8d67d8..3731c58 100644
--- a/sysdeps/powerpc/fpu/e_hypot.c
+++ b/sysdeps/powerpc/fpu/e_hypot.c
@@ -26,7 +26,7 @@ static const double two500  = 3.2733906078961419e+150;
 static const double two600  = 4.149515568880993e+180;
 static const double two1022 = 4.49423283715579e+307;
 static const double twoM500 = 3.054936363499605e-151;
-static const double twoM600 = 4.616489308892868e-128;
+static const double twoM600 = 2.4099198651028841e-181;
 static const double pdnum   = 2.225073858507201e-308;
 
 /* __ieee754_hypot(x,y)
