diff -rup a/sysdeps/ieee754/dbl-64/w_exp.c b/sysdeps/ieee754/dbl-64/w_exp.c
--- a/sysdeps/ieee754/dbl-64/w_exp.c	2012-01-01 05:16:32.000000000 -0700
+++ b/sysdeps/ieee754/dbl-64/w_exp.c	2012-02-24 10:32:52.769230965 -0700
@@ -32,12 +32,12 @@ __exp (double x)
   if (__builtin_expect (x > o_threshold, 0))
     {
       if (_LIB_VERSION != _IEEE_)
-	return __kernel_standard_f (x, x, 6);
+	return __kernel_standard (x, x, 6);
     }
   else if (__builtin_expect (x < u_threshold, 0))
     {
       if (_LIB_VERSION != _IEEE_)
-	return __kernel_standard_f (x, x, 7);
+	return __kernel_standard (x, x, 7);
     }
 
   return __ieee754_exp (x);
