diff --git a/nss/getnssent.c b/nss/getnssent.c
index 014b376..bbcfeb8 100644
--- a/nss/getnssent.c
+++ b/nss/getnssent.c
@@ -32,7 +32,7 @@ __nss_getent (getent_r_function func, void **resbuf, char **buffer,
       *buffer = malloc (*buffer_size);
     }
 
-  while (buffer != NULL
+  while (*buffer != NULL
 	 && func (resbuf, *buffer, *buffer_size, &result, h_errnop) == ERANGE
 	 && (h_errnop == NULL || *h_errnop == NETDB_INTERNAL))
     {
